### ISSUE TITLE
High latency for concurrent LWT updates to the same partition

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#13078]
CQL's LWT write, and Alternator's read-modify-write operations, always read and write a single row, the same row - there are no operations that can do something like read from one row and write to another row. So our LWT implementation should serialize concurrent operations on a single row, but could allow multiple operations to different rows to proceed in parallel.

Unfortunately, Scylla's LWT Paxos implementation currently locks an entire partition while doing an operation on a row, which can cause very long latencies in a workload that does LWT writes concurrently to many different rows in the same partition.

I don't know if Cassandra also locks entire partitions in its LWT implementation (@gleb-cloudius @kostja, do you know ?), but DynamoDB certainly doesn't, which makes Alternator's performance in this use case (many concurrent writes to different items of the same partition) conspicuously bad, and users have complained.

@gleb-cloudius @kostja can you please comment on _why_ a single-row LWT request currently locks an entire partition instead of just one row, and how difficult it would be to make it lock just a single row, and why wasn't this done in the first place?

**Comment by nyh** [comment]
I want to make another comment: When we have two modifications to the same partition, we know that exactly the same RF nodes will handle it, and it is indeed quite wasteful to make several Paxos algorithms in parallel. But the solution shouldn't be to do all these Paxos algorithms one after another instead of in parallel... That doesn't help - it's even worse... 

The ideal solution (I think) is do one Paxos algorithm which does multiple writes to the same partition, but I don't know how easy this is to implement - it requires careful error handling, and somehow collecting the batch of writes from several individual write requests.

**Comment by kostja** [comment]
@nyh in your last comment you describe https://github.com/scylladb/scylladb/issues/6399

**Comment by kostja** [comment]
@nyh regarding your original complaint, if you always work on the row level, what is your reason to have a clustering key at all? Just merge the clustering key into the partition key.

**Comment by nyh** [comment]
> @nyh in your last comment you describe #6399

Right. It seems that if we want to do this issue as _efficiently_ as possible, it becomes a duplicate of #6399.

However, could we have an interim solution for this issue which isn't more efficient than we have today - but just have lower latency? Specifically: Today if the user does 10 concurrent LWT updates to different rows in the same partition, these updates are serialized and we end up with 10 paxos rounds done one after another, with very high latency to the 10th one. What if, instead we did these 10 paxos rounds in parallel? This will still still be inefficient, but at least we'll have lower latencies for the individual writes.

**Comment by kostja** [comment]
> > @nyh regarding your original complaint, if you always work on the row level, what is your reason to have a clustering key at all? Just merge the clustering key into the partition key.
> 
> Well, in any case, in writes indeed "you always work on the row level"... The reason to have a clustering key has to do with reads, not writes. Namely, clustering keys are useful if you want to efficiently read a sorted list of items with the same partition key. This need is orthogonal to the use of LWT in the writes.

We could I suppose have a table option to mae

> > @nyh in your last comment you describe #6399
> 
> Right. It seems that if we want to do this issue as _efficiently_ as possible, it becomes a duplicate of #6399.
> 
> However, could we have an interim solution for this issue which isn't more efficient than we have today - but just have lower latency? Specifically: Today if the user does 10 concurrent LWT updates to different rows in the same partition, these updates are serialized and we end up with 10 paxos rounds done one after another, with very high latency to the 10th one. What if, instead we did these 10 paxos rounds in parallel? This will still still be inefficient, but at least we'll have lower latencies for the individual writes.

Generally speaking no, since we can't know if we won't have LWT batches which use the entire partition and for which we promise linearizability. There is also a guarantee/expectation of linearizing LWT writes to static columns. See, LWT is tightly coupled with Cassandra data model.  I think linearizing batches would be an easier fix than allowing what you suggest.

**Comment by nyh** [comment]
> Generally speaking no, since we can't know if we won't have LWT batches which use the entire partition and for which we promise linearizability. There is also a guarantee/expectation of linearizing LWT writes to static columns.

This could be solved exactly like we did it in materialized views (where you can also have operations on a single row or the entire partition) by having a two-level lock where you can lock just a row (which locks out both operations on the row and operations on the entire partition) or an entire partition. Maybe we could even use the same implementation in db/view/row_locking.hh.

By the way, in Alternator (which was the original motivation for this issue), we don't have whole-partition operations nor static columns. So we consider having a quick-and-dirty implementation just for Alternator tables (although of course it would be even better to have a perfect solution for all use cases).

> See, LWT is tightly coupled with Cassandra data model. I think coalescing a queue of writes into a single batch would be an easier fix than allowing what you suggest.

**Comment by nyh** [comment]
> Let's assume for a moment that your proposal should be preferred instead of #6399, what do you think are the next steps?

My idea for a workaround is not "preferred" - it is worse than your idea - I just thought it might be easier as an ad-interim solution, and also require less LWT expertise to implement. My workaround proposal will not improve throughput at all but will improve the latency problem that users complained about.

**Comment by calvin2021y** [comment]
after tablet + raft, this will resolved?

**Comment by nyh** [comment]
> after tablet + raft, this will resolved?

No. Or rather, it depends what you mean by "tablet + raft". The code which is _currently_ being developed for tablets has a lot of great benefits, but does nothing to improve LWT performance or solve this issue. In the future, we are likely to develop a new version of consistent tables (with builtin support for LWT) based on Raft, but this is a future idea - it is not being worked on today.

**Comment by nyh** [comment]
> However, could we have an interim solution for this issue which isn't more efficient than we have today - but just have lower latency? Specifically: Today if the user does 10 concurrent LWT updates to different rows in the same partition, these updates are serialized and we end up with 10 paxos rounds done one after another, with very high latency to the 10th one. What if, instead we did these 10 paxos rounds in parallel? This will still still be inefficient, but at least we'll have lower latencies for the individual writes.

@gleb-cloudius explained to me that the per-partition lock isn't the whole story - there is also a paxos table whose key is the table and partition-key (_not_ a clustering key) and finds the current ballot for this key. So if we want to allow several concurrent ballots in the same partition, we'll need to change the layout of the paxos table :-(

**Comment by avikivity** [comment]
Not easy to fix. We'll need a paxos_v2 table and to support both (concurrently!) for a transition period.

**Comment by gleb-cloudius** [comment]
I am also not sure paxos table can be made per ck since IIRC we support multi ck transactions with lwt. @kostja do we?

**Comment by kostja** [comment]
> I am also not sure paxos table can be made per ck since IIRC we support multi ck transactions with lwt. @kostja do we?

Yes, we do, and there is an alternative part forward already described in our issue database and implemented in Cassandra 5 - it's https://github.com/scylladb/scylladb/issues/6399
